Revise the modules document for clarity

AaronBallman opened the pull request:
The intention isn't to add or change the information provided, but to improve clarity through some grammar fixes, improvements to the markdown, and so forth.

AaronBallman committed:
Revise the modules document for clarity

The intention isn't to add or change the information provided, but to
improve clarity through some grammar fixes, improvements to the
markdown, and so forth.

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -444,53 +435,57 @@ For example, the following example is allowed:
-
-Only Itanium C++ ABI related change are mentioned
+This section describes the new ABI changes brought by modules. Only Itanium C++
+ABI related change are mentioned.
```
"Only changes to Itanium C++ ABI are covered."

erichkeane replied:
```suggestion
ABI related changes are mentioned.
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -633,36 +631,36 @@ example:
+In the above example, the function definition of ``N::g`` is elided from the
+Reduced BMI of ``M.cppm``. Then the use of ``use_g<int>`` in ``M-impl.cpp``
+fails to instantiate. For such issues, users can add references to ``N::g`` in
+the module purview of ``M.cppm`` to ensure it is reachable, e.g.,
```
```suggestion
the module purview of ``M.cppm`` to ensure it is reachable, e.g.
```

erichkeane replied:
what is 'module purview' here?  Do we just mean "in the `M` module"?

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -633,36 +631,36 @@ example:
+``using N::g;``.
+
+Long-term, Clang is likely to make Reduced BMIs the default rather than Full
+BMIs. But because it is a drastic change, it is initially an experimental
```
```suggestion
BMIs. Since it would be a drastic change of user interface, it is initially an experimental
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -633,36 +631,36 @@ example:
+   ``-fmodules-reduced-bmi`` as a new option, and recommend use of the new
+   option. This transition is expected to take 1~2 additional releases as well.
+3. Finally, ``-fmodules-reduced-bmi`` will be the default. When that time
+   comes, the term BMI will refer to the reduced BMI and the Full BMI will only
```
```suggestion
   comes, the term BMI will refer to the Reduced BMI and the Full BMI will only
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -633,36 +631,36 @@ example:
+While it is valid to have duplicated declarations in the global module fragments
+of different module units, it is not free for Clang to deal with the duplicated
+declarations. A translation unit will compile more slowly if it and its
+imported module units contain a lot of duplicated declarations. For example:
```
"A translation unit will compile more slowly if there is a lot of duplicated declarations between the translation unit and modules it imports."

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -738,22 +736,21 @@ the following style significantly:
-
-For many existing libraries, it may be a breaking change to refactor themselves
-into modules completely. So that many existing libraries need to provide headers and module
+New code and libraries should use modules from day one if possible. However,
```
```suggestion
New code and libraries should use modules from day 1 if possible. However,
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -738,22 +736,21 @@ the following style significantly:
-For many existing libraries, it may be a breaking change to refactor themselves
-into modules completely. So that many existing libraries need to provide headers and module
+New code and libraries should use modules from day one if possible. However,
+it may be a breaking change for existing code or libraries to refactor to use
```
```suggestion
it may be a breaking change for existing code or libraries to switch to
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -738,22 +736,21 @@ the following style significantly:
-into modules completely. So that many existing libraries need to provide headers and module
+New code and libraries should use modules from day one if possible. However,
+it may be a breaking change for existing code or libraries to refactor to use
+modules. As a result, many existing libraries need to provide headers and module
```
```suggestion
modules. As a result, many existing libraries need to provide both headers and module
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -738,22 +736,21 @@ the following style significantly:
 
-Let's start with the case that there is no dependency or no dependent libraries providing
-modules for your library.
+This section provides some ideas on how to ease the transition process for
```
```suggestion
This section suggests some ideas on how to ease the transition process for
```

erichkeane replied:
```suggestion
This section provides some suggestions on how to ease the transition process for
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -776,9 +773,9 @@ export-using style
-to be exported and `using` such declarations in an `export` block. Then, basically,
-we're done.
+This example shows how to include all the headers containing declarations which
+need to be exported and uses `using` declarations in an `export` block to
```
```suggestion
need to be exported, and uses `using` declarations in an `export` block to
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -776,9 +773,9 @@ export-using style
-we're done.
+This example shows how to include all the headers containing declarations which
+need to be exported and uses `using` declarations in an `export` block to
+produce the module interface. Then, basically, we're done.
```
```suggestion
produce the module interface.
```

erichkeane replied:
```suggestion
produce the module interface. 
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -900,16 +897,17 @@ new ABI format. This is the job of the additional part of the interface unit:
-when the number of the source files are not small enough.
+If the number of source files is small, everything can be put in the private
+module fragment directly. (It is recommended to add conditional includes to the
+source files too). However, compile time performance will be slow if there are
```
```suggestion
source files too). However, compile time performance will be bad if there is
```

erichkeane replied:
```suggestion
source files as well). However, compile time performance will be slow if there are
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -925,45 +923,41 @@ In that case, you need to convert your source files (.cpp files) to module imple
-
-If you're a perfectionist, maybe you can improve it by introducing internal module partition unit.
+The above practice may be problematic if there are headers only included by the
+source files. If using a private module fragment, this issue may be solved by
```
```suggestion
source files. When using a private module fragment, this issue may be solved by
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1069,35 +1061,39 @@ For ABI breaking style,
-We don't need the non-exported using declarations if we're using implementation module
-units now. We can import thirdparty modules directly in the implementation module
-units.
+Non-exported using declarations are not needed if using implementation module
```
```suggestion
Non-exported ``using`` declarations are not needed if using implementation module
```

erichkeane replied:
```suggestion
Non-exported using declarations are unnecessary if using implementation module
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -444,53 +435,57 @@ For example, the following example is allowed:
+Object definition consistency
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The C++ language requires that the same declarations in different translation
```
```suggestion
The C++ language requires that the declarations of the same entity in different translation
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1126,33 +1122,31 @@ but it will get rejected if we reverse the order of ``#include <iostream>`` and
-
-Ignored PreferredName Attribute
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This is a limitation in the implementation. In the first example, the compiler
```
```suggestion
This is a limitation of the implementation. In the first example, the compiler
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1126,33 +1122,31 @@ but it will get rejected if we reverse the order of ``#include <iostream>`` and
 
-Due to a tricky problem, when Clang writes BMIs, Clang will ignore the ``preferred_name`` attribute, if any.
-This implies that the ``preferred_name`` wouldn't show in debugger or dumping.
+Ignored ``prefered_name`` Attribute
```
```suggestion
Ignored ``preferred_name`` Attribute
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1222,33 +1214,31 @@ One often reported example is:
-This is tracked in https://github.com/llvm/llvm-project/issues/78850.
+Currently the compiler incorrectly diagnoses the inconsistent definition of
+``fun()`` in two module units. Because both definitions of ``fun()`` have the
+same spelling and ``T`` refers to the same type entity, so there is no ODR
```
```suggestion
same spelling and ``T`` refers to the same type entity, there is no ODR
```

erichkeane replied:
```suggestion
same spelling and ``T`` refers to the same type entity, there is no ODR
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1414,46 +1404,43 @@ In fact, we could even "mimic" the sytle of header units by Clang modules:
-differences now would likely become a problem in the future.
+Another reason is that there are proposals to introduce module mappers to the
+C++ standard (for example, https://wg21.link/p1184r2). Reusing Clang's
+modulemap may be more difficult if we need to introduce another module mapper.
```
```suggestion
``modulemap`` may be more difficult if we need to introduce another module mapper.
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1785,18 +1773,17 @@ First is ``O0``. The compilation process is described in the following graph.
-But with optimizations, things are different:
+In this case, the source file (which could be a non-module unit or a module
+unit) would get processed by the whole pipeline. But the imported code would
+only get involved in semantic analysis, which is mainly about name lookup,
```
```suggestion
only get involved in semantic analysis, which, for the most part, is name lookup,
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1817,27 +1804,29 @@ But with optimizations, things are different:
+It would be very unfortunate if we end up with worse performance when using
+modules. The main concern is that when a source file is compiled, the compiler
+needs to see the body of imported module units so that it can perform IPO
+(InterProcedural Optimization, primarily inlining in practice) to optimize
```
```suggestion
(Interprocedural Optimization, primarily inlining in practice) to optimize
```

Endilll commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1817,27 +1804,29 @@ But with optimizations, things are different:
+not impact module users, but at higher optimization levels the definitions of
+such functions are provided to user compilations for the purposes of
+optimization (but definitions of these functions are still not included in the
+use's object file) -- this means the build speedup at higher optimization
```
```suggestion
use's object file). This means the build speedup at higher optimization
```

Endilll reviewed (commented):
I have to admit that is rather shallow due to sheer amount of text, but it still took me more like 90 minutes. Hope you find it useful.

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -400,24 +389,27 @@ And the compilation process for module units are like:
-If we envision modules as a cache to speed up compilation, then - as with other caching techniques -
-it is important to keep cache consistency.
-So **currently** Clang will do very strict check for consistency.
+If modules are thought of as a kind of cache to speed up compilation, then, as
```
Modules are effectively a cache used to speed up compilation.  Thus, like other caching techniques, it is important to maintain cache conistency.  Therefore, Clang does very strictly checking for consistency.

AaronBallman replied:
I went with:

Modules can be viewed as a kind of cache to speed up compilation. Thus, like
other caching techniques, it is important to maintain cache consistency which
is why Clang does very strict checking for consistency.

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -444,53 +435,57 @@ For example, the following example is allowed:
+
+The C++ language requires that the same declarations in different translation
+units have the same definition, which is known as the One Definition Rule (ODR).
+Prior to modules, the compiler cannot perform strong ODR violation checking
```
```suggestion
Without modules, the compiler cannot perform strong ODR violation checking
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -444,53 +435,57 @@ For example, the following example is allowed:
+The C++ language requires that the same declarations in different translation
+units have the same definition, which is known as the One Definition Rule (ODR).
+Prior to modules, the compiler cannot perform strong ODR violation checking
+because it only sees one translation unit at a time. With use of modules, the
```
```suggestion
because it only sees one translation unit at a time. With the use of modules, the
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -530,43 +527,43 @@ Now the linkage name of ``NS::foo()`` will be ``_ZN2NS3fooEv``.
-at the point of appearance of the import declaration.
+All importable module units are required to emit an initializer function. The
+initializer function emits calls to imported modules first followed by calls
+to all to dynamic initializers in the current module unit.
```
```suggestion
to all of the dynamic initializers in the current module unit.
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -530,43 +527,43 @@ Now the linkage name of ``NS::foo()`` will be ``_ZN2NS3fooEv``.
-the dynamic-initializers in the TU. Initializations of entities at namespace
-scope are appearance-ordered. This (recursively) extends into imported modules
-at the point of appearance of the import declaration.
+All importable module units are required to emit an initializer function. The
```
```suggestion
All importable module units are required to emit an initializer function to handle <what?>. The
```

ARE they required, or only if there is some sort of static initialization happening in the module?

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -530,43 +527,43 @@ Now the linkage name of ``NS::foo()`` will be ``_ZN2NS3fooEv``.
+the initializer functions of the imported named module within the sequence of
+the dynamic initializers in the translation unit. Initializations of entities
+at namespace scope are appearance-ordered. This (recursively) extends to
+imported modules at the point of appearance of the import declaration.
```
Do we want to clarify that in the case of template instantiations (such as a variable template with an initializer) that this 'order of appearance' isn't quite true anymore?

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -530,43 +527,43 @@ Now the linkage name of ``NS::foo()`` will be ``_ZN2NS3fooEv``.
-contained in the BMI.
+To support the two-phase compilation model, Clang puts everything needed to
+produce an object into the BMI. However, other consumers of the BMI generally
+don't need that informations. This makes the BMI larger and may introduce
```
```suggestion
don't need that information. This makes the BMI larger and may introduce
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -577,15 +574,16 @@ the generated BMI specified by ``-o`` will be full BMI and the BMI specified by
-early without asking for the help of build systems. The users of build systems which supports
-two phase compilation model still need helps from build systems.
+Clang does not emit diagnostics when ``-fexperimental-modules-reduced-bmi`` is
+used with a non-module unit. This design helps the end users of the one-phase
```
```suggestion
used with a non-module unit. This design permits users of the one-phase
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -738,22 +736,21 @@ the following style significantly:
-Let's start with the case that there is no dependency or no dependent libraries providing
-modules for your library.
+This section provides some ideas on how to ease the transition process for
+existing libraries. **Note that this information is only intended as helpful
```
```suggestion
existing libraries. **Note that this information is only intended as guidance, rather than as requirements to use modules in Clang.
```

cor3ntin commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -8,79 +8,60 @@ Standard C++ Modules
-either "Named Modules" or to "Header Units". This document covers both forms.
+The term ``modules`` has a lot of meanings. For Clang users, modules may refer
+to ``Objective-C Modules``, `Clang Modules <Modules.html>`_ (also called
+``Clang Header Modules``, etc.) or ``C++20 Modules`` (or
```
Etc?

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -875,11 +872,11 @@ The pattern for ABI breaking style is similar with export extern-C++ style.
 
-Remember that the ABI get changed and we need to compile our source files into the
-new ABI format. This is the job of the additional part of the interface unit:
+The ABI with modules is different and so we need to compile the source files
```
```suggestion
The ABI with modules is different and thus we need to compile the source files
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -900,16 +897,17 @@ new ABI format. This is the job of the additional part of the interface unit:
-files too). But it will make the compilation of the module interface unit to be slow
-when the number of the source files are not small enough.
+If the number of source files is small, everything can be put in the private
+module fragment directly. (It is recommended to add conditional includes to the
```
```suggestion
module fragment directly (It is recommended to add conditional includes to the
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -925,45 +923,41 @@ In that case, you need to convert your source files (.cpp files) to module imple
-This is the reason why we add non-exported using declarations from the third
-party libraries in the primary module interface unit.
+The module implementation unit will import the primary module implicitly. Do
+not include any headers in the module implementation units because that avoids
```
```suggestion
not include any headers in the module implementation units as it avoids
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -925,45 +923,41 @@ In that case, you need to convert your source files (.cpp files) to module imple
+The module implementation unit will import the primary module implicitly. Do
+not include any headers in the module implementation units because that avoids
+duplicated declarations between translation units. This is why non-exported
+using declarations are added from third-party libraries in the primary module
```
```suggestion
using declarations should be added from third-party libraries in the primary module
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -998,25 +992,23 @@ headers to:
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Life gets easier if all the dependent libraries providing modules.
+Life gets easier if all the dependent libraries provide modules.
```
```suggestion
Of course, most of the complexity disappears if all the dependent libraries provide modules.
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1069,35 +1061,39 @@ For ABI breaking style,
-headers, we should include that after the including. If the imported modules don't provide
-the headers, we can make it ourselves if we still want to optimize it.
+If the library has to mix the use of ``include`` and ``import`` in its module,
+the key point is still to remove duplicated declarations in translation units
```
```suggestion
the primary goal is still the removal of duplicated declarations in translation units
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1069,35 +1061,39 @@ For ABI breaking style,
-and add the label ``clang:modules`` (if you have permissions for that).
+The following describes issues in the current implementation of modules. Please
+see
+`the issues list for moduled <https://github.com/llvm/llvm-project/labels/clang%3Amodules>`_
```
```suggestion
`the issues list for modules <https://github.com/llvm/llvm-project/labels/clang%3Amodules>`_
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1162,39 +1156,37 @@ A direct idea would be use macros like:
-once P1857R3 is implemented.
-A simple suggestion would be "Don't play macro tricks with module declarations".
+With the idea being that this file can be compiled like a module unit or a
+non-module unit depending on the definition of some macros. However, this kind
```
```suggestion
non-module unit depending on the definition of some macros. However, this
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1162,39 +1156,37 @@ A direct idea would be use macros like:
-A simple suggestion would be "Don't play macro tricks with module declarations".
+With the idea being that this file can be compiled like a module unit or a
+non-module unit depending on the definition of some macros. However, this kind
+of usage is forbidden by P1857R3 which is not yet implemented in Clang. This
```
```suggestion
usage is forbidden by P1857R3 which is not yet implemented in Clang. This
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1162,39 +1156,37 @@ A direct idea would be use macros like:
-This means that is possible to write illegal modules code now, and obviously this will stop working
-once P1857R3 is implemented.
-A simple suggestion would be "Don't play macro tricks with module declarations".
+With the idea being that this file can be compiled like a module unit or a
```
```suggestion
The intent of this is that this file can be compiled like a module unit or a
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1162,39 +1156,37 @@ A direct idea would be use macros like:
+With the idea being that this file can be compiled like a module unit or a
+non-module unit depending on the definition of some macros. However, this kind
+of usage is forbidden by P1857R3 which is not yet implemented in Clang. This
+means that is possible to write invalid modules which no longer be accepted
```
```suggestion
means that is possible to write invalid modules which will no longer be accepted
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1222,33 +1214,31 @@ One often reported example is:
-This results in unclear and confusing diagnostic messages.
-And it is worse that the compiler may import TU-local entities to other units without any
-diagnostics.
+However, Clang doesn't formally support these two ideas. This results in
```
```suggestion
However, Clang doesn't formally support these two concepts. This results in
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1222,33 +1214,31 @@ One often reported example is:
-And it is worse that the compiler may import TU-local entities to other units without any
-diagnostics.
+However, Clang doesn't formally support these two ideas. This results in
+unclear or confusing diagnostic messages. Further, Clang may import TU-local
```
```suggestion
unclear or confusing diagnostic messages. Further, Clang may import ``TU-local``
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1335,19 +1324,22 @@ For example,
 ~~~~~~~~~~~~~~~~~~~~~
 
-Another difference with modules is that we can't compile the BMI from a header unit.
+Because of the semantics of header units, which are just like headers, another
```
Oh my...
```suggestion
With modules, we cannot compile the BMI from a header unit due to the semantics of header units.
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1391,17 +1381,17 @@ with the following one:
-
-In fact, we could even "mimic" the sytle of header units by Clang modules:
+Header units have similar semantics to Clang modules. The semantics of both are
+like headers. In fact, header units can be mimicked by Clang modules as in the
```
```suggestion
like headers. Therefore, header units can be mimicked by Clang modules as in the
```

erichkeane replied:
Do we really mean 'Clang Modules' here?  I thought we were referring to C++ Standards Modules throughout here?

AaronBallman replied:
I think we do mean Clang modules here, this is in a section about the relationship between C++ modules and Clang modules.

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1414,46 +1404,43 @@ In fact, we could even "mimic" the sytle of header units by Clang modules:
-Another reason is that there are proposals to introduce module mappers to the C++ standard
-(for example, https://wg21.link/p1184r2).
-If we decide to reuse Clang's modulemap, we may get in trouble once we need to introduce another module mapper.
+The main reason is because Clang modules have more hierarchical semantics when
```
```suggestion
This is primarily because Clang modules have more hierarchical semantics when
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1414,46 +1404,43 @@ In fact, we could even "mimic" the sytle of header units by Clang modules:
 
-The clang-scan-deps scanner implemented
+The ``clang-scan-deps`` scanner implemented
 `P1689 paper <https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1689r5.html>`_
```
```suggestion
`P1689 <https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1689r5.html>`_
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1753,17 +1740,18 @@ Possible Questions
-In a simpler word, we could get rid of many redundant compilations by using modules.
+A classic theory for the reason why modules speed up the compilation is: if
+there are ``n`` headers and ``m`` source files and each header is included by
+each source file, then the complexity of the compilation is ``O(n*m)``. But if
```
```suggestion
each source file, then the complexity of the compilation is ``O(n*m)``. However if
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1753,17 +1740,18 @@ Possible Questions
+there are ``n`` headers and ``m`` source files and each header is included by
+each source file, then the complexity of the compilation is ``O(n*m)``. But if
+there are ``n`` module interfaces and ``m`` source files, the complexity of the
+compilation is ``O(n+m)``. So, using modules would be a big win when scaling.
```
```suggestion
compilation is ``O(n+m)``. So, using modules would be a significant improvement at scale.
```

cor3ntin commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -400,24 +389,27 @@ And the compilation process for module units are like:
-So **currently** Clang will do very strict check for consistency.
+If modules are thought of as a kind of cache to speed up compilation, then, as
+with other caching techniques, it is important to keep cache consistency.
+**Currently**, Clang does very strict checking for consistency.
```
The "currently" is not useful here

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1753,17 +1740,18 @@ Possible Questions
+each source file, then the complexity of the compilation is ``O(n*m)``. But if
+there are ``n`` module interfaces and ``m`` source files, the complexity of the
+compilation is ``O(n+m)``. So, using modules would be a big win when scaling.
+In a simpler word, we could get rid of many redundant compilations by using
```
```suggestion
More simply, modules causes much of the redundant compilations to no longer be necessary.
```

cor3ntin commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -444,53 +435,57 @@ For example, the following example is allowed:
+definitions. Further, true positive ODR violations are not always reported.
+Additionally, MSVC does not currently perform ODR checking for declarations in
+the global module fragment.
+
```
Why are we talking about msvc? The mention below is sufficient

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1785,18 +1773,17 @@ First is ``O0``. The compilation process is described in the following graph.
-
-But with optimizations, things are different:
+In this case, the source file (which could be a non-module unit or a module
+unit) would get processed by the whole pipeline. But the imported code would
```
```suggestion
unit) would get processed by the entire pipeline. However, the imported code would
```

erichkeane commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1817,27 +1804,29 @@ But with optimizations, things are different:
+processed again and again in importee units by optimizations (including IPO
+itself). The optimizations before IPO and IPO itself are the most time-consuming
+part in whole compilation process. So from this perspective, it might not be
+possible to get the compile time improvements described in the theory, but
```
```suggestion
possible to get the compile time improvements described, but
```

cor3ntin commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -530,43 +527,43 @@ Now the linkage name of ``NS::foo()`` will be ``_ZN2NS3fooEv``.
+To support the two-phase compilation model, Clang puts everything needed to
+produce an object into the BMI. However, other consumers of the BMI generally
+don't need that informations. This makes the BMI larger and may introduce
+unnecessary dependencies for the BMI. To mitigate the problem, Clang added a
```
added is weird. It should be more declarative/present tense.

Offers/has

cor3ntin commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -530,43 +527,43 @@ Now the linkage name of ``NS::foo()`` will be ``_ZN2NS3fooEv``.
-To be clear, we call the default BMI as Full BMI and the new introduced BMI as Reduced
-BMI.
+Users can use the ``-fexperimental-modules-reduced-bmi`` option to produce a
+Reduced BMI.
```
We should find a better name in a subsequent pr

cor3ntin commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -577,15 +574,16 @@ the generated BMI specified by ``-o`` will be full BMI and the BMI specified by
-two phase compilation model still need helps from build systems.
+Clang does not emit diagnostics when ``-fexperimental-modules-reduced-bmi`` is
+used with a non-module unit. This design helps the end users of the one-phase
+compilation model to perform experiments without needing to modify the build
```
Perform experiments?

cor3ntin commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -577,15 +574,16 @@ the generated BMI specified by ``-o`` will be full BMI and the BMI specified by
+module fragment, or definitions of non-inline functions and non-inline
+variables. This may not be a transparent change.
+
+[module.global.frag]ex2 provides a good example:
```
The specific c++ standard reference isn't useful. Just "consider the following example"

cor3ntin commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -633,36 +631,36 @@ example:
+the module purview of ``M.cppm`` to ensure it is reachable, e.g.,
+``using N::g;``.
+
+Long-term, Clang is likely to make Reduced BMIs the default rather than Full
```
Support for reduced BMI is still experimental but it might become the default in the future.

AaronBallman committed:
Updated based on review feedback

This got most of the comments from Vlad and Corentin, but it does not
address comments from Erich.

Additionally, it switches the file to UTF-8 from ANSI Latin 1.

ChuanqiXu9 reviewed (commented):
Big thanks!

I left some comments about correctness or clearness. And all other change looks good to me.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-modules-documentation

AaronBallman commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -8,79 +8,60 @@ Standard C++ Modules
-translation unit. Every module unit must have a module declaration. The syntax
-of the module declaration is:
+A module consists of one or more module units. A module unit is a special kind
+of translation unit. Every module unit must have a module declaration. The
```
I changed it to "A module unit should almost always start with a module declaration." as that seems sufficiently hand-wavy.

AaronBallman commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -312,75 +300,76 @@ So all of the following name is not valid by default:
 
-``-fprebuilt-module-path`` is more convenient and ``-fmodule-file`` is faster since
-it saves time for file lookup.
+``-fprebuilt-module-path`` is more convenient while ``-fmodule-file`` is faster
```
I've removed it.

AaronBallman commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -8,109 +8,91 @@ Standard C++ Modules
 
-Built Module Interface file
-~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Binary Module Interface
```
I don't understand the distinction between Binary Module Interface and Built Module Interface file -- from a user perspective, they're saying the same thing: it's a non-textual (binary) representation of a module interface.

ChuanqiXu9 replied:
Maybe it is slightly `pedantic`. The term `Binary Module Interface` may make people wondering this is similar with `ABI`, which is a binary representation and an interface. But now the BMI is not like ABI since the compiler can't consume the BMI from different compilers even if they are from different versions.

But maybe this is not a big deal really.

AaronBallman commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -400,24 +389,27 @@ And the compilation process for module units are like:
 
-The language option of module units and their non-module-unit users should be consistent.
-The following example is not allowed:
+Language dialect compiler options for module units and their non-module-unit
```
How about I go with "Compiler options related to the language dialect for a module unit and its non-module-unit uses need to be consistent."?

The problem is "language option" isn't sufficient, right? e.g., enabling `-fopenmp` in one place and not in the other would be just as much of a problem as setting `-std=c++20` in one place and not in the other.

AaronBallman commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -738,22 +736,21 @@ the following style significantly:
 
-For new libraries, we encourage them to use modules completely from day one if possible.
-This will be pretty helpful to make the whole ecosystems to get ready.
+New code and libraries should use modules from the start if possible. However,
```
I reworded slightly to "It is best for new code and libraries to use modules from the start if possible." -- do you think that's an improvement?

ChuanqiXu9 replied:
Yeah, I think so.

AaronBallman commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -925,45 +923,41 @@ In that case, you need to convert your source files (.cpp files) to module imple
-But it may not be too hard if you can understand the points of the section.
+This can potentially be improved by introducing module partition implementation
+unit. The module partition implementation unit is an importable module unit
+which is internal to the module itself. However, this approach may not always
+be the best way forward.
```
Ah, I read the original wording wrong, I'll drop this sentence entirely.

AaronBallman committed:
Update based on review feedback

AaronBallman commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -8,109 +8,91 @@ Standard C++ Modules
+* Primary module interface unit
+* Module implementation unit
+* Module partition interface unit
+* Module partition implementation unit
```
Okay, I've reverted back to the old form.

AaronBallman committed:
Module partition implementation unit -> Internal module partition unit

AaronBallman commented:
I think I've addressed all of the feedback (thank you all for the improvements!), except for the suggestion from @Endilll to switch to using italics rather than code font. I think the suggestion is a good one, but it's a time-consuming edit that doesn't change the meaning of the document, so I'd prefer that be done in a follow-up.

ChuanqiXu9 reviewed (approved):
LGTM, thanks : )

AaronBallman committed:
Remove trailing whitespace; NFC

AaronBallman committed:
Binary Module Interface -> Built Module Interface

AaronBallman committed:
Updates based on offline feedback from Corentin

whisperity commented on clang/docs/StandardCPlusPlusModules.rst:
```diff
@@ -1753,91 +1742,93 @@ Possible Questions
-Please file new github issues as you find interoperability problems.
+We **wish** to support Clang modules and standard C++ modules at the same time,
+but the mixing them together is not well used/tested yet. Please file new
+github issues as you find interoperability problems.
```
```suggestion
GitHub issues as you find interoperability problems.
```

AaronBallman replied:
Good catch! Done.

AaronBallman committed:
github -> GitHub
